ui/src/main/java/edu/wpi/grip/ui/preview/PreviewsView.java

```diff
@@ -48,24 +58,196 @@ public PreviewsView(EventBus eventBus) {
+            while (previewedMovedSockets.size() != 0) { //While there are still sockets to deal with on the stack
+                OutputSocket<?> current = previewedMovedSockets.pop();//Grab the top socket on the stack
+                int oldIndex = this.previewedSockets.indexOf(current);//Get the index of this preview so we can remove the correct entry
+
+                int newLocation = 0;//This will hold the new index in the list of previewed sockets for this socket
+
+                if (distanceMoved < 0) //If the step moved left....
+                    newLocation = leftmostIndex + distanceMoved; //Calculate the new index from the leftmost previewed socket of this step
+                else //The step must have moved right....
+                    newLocation = rightmostIndex + distanceMoved;//So calculate the new index from the rightmost previewed socket of this step
+
+                if (newLocation < numberOfSourcePreviews) {//If the new calculated index would put it in the midst of source previews
+                    newLocation = numberOfSourcePreviews;//Make the index the location of the first non-source preview
+                } else { //The new index is the current location of another step (NOT a source)
+
+                    //So we need to make sure that we jump over GROUPS of previews associated with the SAME step as a unit
+                    int count = 0;//This will hold the number of previews open from the same step in sequence, starting from the new location and going in the direction we are moving
+
+                    if (distanceMoved < 0) {//If the step moved left....
+                        OutputSocket<?> nextSocketInDirection = this.previewedSockets.get(newLocation);//Grab the socket whose preview is open at the new location
+                        boolean zeroReached = false;//We will set this to true if we reach the beginning of the list of previews (there are no source previews open)
+                        while ((!zeroReached) &&
+                                ((nextSocketInDirection.getStep().isPresent())
+                                        && (nextSocketInDirection.getStep().get() == this.previewedSockets.get(newLocation).getStep().get()))) { //While we haven't reached the beginning of the list of previews, the socket at this location is a socket from a step, and it is the SAME step as the step of the socket at the new location...
+                            count++;
+                            if ((newLocation - count) > 0)//If we haven't reached the beginning of the list of open previews...
+                                nextSocketInDirection = this.previewedSockets.get(newLocation - count);//Grab the next previewed socket to examine in the direction we are moving
+                            else
+                                zeroReached = true;//Mark that we've reached the beginning of the list of previews so we know to stop looking for more
```

Comment:
So as a general practice always wrap your if statement code in blocks `{ ... }` OR put it on the same line
`else zeroReached = true;`

I enforce this because merging code can create bugs that are really hard to spot.
See the major apple security bug that was caused by having two gotos after an if statement: https://nakedsecurity.sophos.com/2014/02/24/anatomy-of-a-goto-fail-apples-ssl-bug-explained-plus-an-unofficial-patch/